model/histogram/float_histogram.go

```diff
@@ -305,14 +305,164 @@ func addBucket(
 
 // Compact eliminates empty buckets at the beginning and end of each span, then
 // merges spans that are consecutive or at most maxEmptyBuckets apart, and
-// finally splits spans that contain more than maxEmptyBuckets. The compaction
-// happens "in place" in the receiving histogram, but a pointer to it is
-// returned for convenience.
+// finally splits spans that contain more consecutive empty buckets than
+// maxEmptyBuckets. (The actual implementation might do something more efficient
+// but with the same result.)  The compaction happens "in place" in the
+// receiving histogram, but a pointer to it is returned for convenience.
 func (h *FloatHistogram) Compact(maxEmptyBuckets int) *FloatHistogram {
-	// TODO(beorn7): Implement.
+	h.PositiveBuckets, h.PositiveSpans = compactBuckets(
+		h.PositiveBuckets, h.PositiveSpans, maxEmptyBuckets,
+	)
+	h.NegativeBuckets, h.NegativeSpans = compactBuckets(
+		h.NegativeBuckets, h.NegativeSpans, maxEmptyBuckets,
+	)
 	return h
 }
 
+func compactBuckets(buckets []float64, spans []Span, maxEmptyBuckets int) ([]float64, []Span) {
+	if len(buckets) == 0 {
+		return buckets, spans
+	}
+
+	var iBucket, iSpan int
+	var posInSpan uint32
+
+	// Helper function.
+	emptyBucketsHere := func() int {
+		i := 0
+		for i+iBucket < len(buckets) &&
+			uint32(i)+posInSpan < spans[iSpan].Length &&
+			buckets[i+iBucket] == 0 {
+			i++
+		}
+		return i
+	}
+
+	// Merge spans with zero-offset to avoid special cases later.
+	if len(spans) > 1 {
+		for i, span := range spans[1:] {
+			if span.Offset == 0 {
+				spans[iSpan].Length += span.Length
+				continue
+			}
+			iSpan++
+			if i+1 != iSpan {
+				spans[iSpan] = span
+			}
+		}
+		spans = spans[:iSpan+1]
+		iSpan = 0
+	}
+
+	// Merge spans with zero-length to avoid special cases later.
+	for i, span := range spans {
+		if span.Length == 0 {
+			if iSpan+1 < len(spans) {
+				spans[iSpan+1].Offset += span.Offset
```

Comment:
Should it be `i+1` instead of `iSpan+1`?

Reply:
If yes, then it needs an unit test.

Reply:
Confirmed that it should be `i+1` with this test case
```go
		{
			"eliminate zero length 2",
			&FloatHistogram{
				PositiveSpans:   []Span{{-2, 2}, {2, 0}, {2, 0}, {2, 0}, {3, 3}},
				PositiveBuckets: []float64{1, 3.3, 4.2, 0.1, 3.3},
			},
			0,
			&FloatHistogram{
				PositiveSpans:   []Span{{-2, 2}, {9, 3}},
				PositiveBuckets: []float64{1, 3.3, 4.2, 0.1, 3.3},
			},
		},
```

Reply:
Great catch. Added fix and test.